cmd/containerboot/main.go

```diff
@@ -0,0 +1,426 @@
+			}
+		}
+	}
+
+	// Context is used for all setup stuff until we're in steady
+	// state, so that if something is hanging we eventually time out
+	// and crashloop the container.
+	ctx, cancel := context.WithTimeout(context.Background(), 60*time.Second)
+	defer cancel()
+
+	if cfg.InKubernetes && cfg.KubeSecret != "" && cfg.AuthKey == "" {
+		key, err := findKeyInKubeSecret(ctx, cfg.KubeSecret)
+		if err != nil {
+			log.Fatalf("Getting authkey from kube secret: %v", err)
+		}
+		if key != "" {
+			log.Print("Using authkey found in kube secret")
+			cfg.AuthKey = key
+		} else {
+			log.Print("No authkey found in kube secret and TS_AUTHKEY not provided, login will be interactive if needed.")
+		}
+	}
+
+	st, err := startAndAuthTailscaled(ctx, cfg)
+	if err != nil {
+		log.Fatalf("failed to bring up tailscale: %v", err)
+	}
+
+	if cfg.ProxyTo != "" {
+		if err := installIPTablesRule(ctx, cfg.ProxyTo, st.TailscaleIPs); err != nil {
+			log.Fatalf("installing proxy rules: %v", err)
+		}
+	}
+	if cfg.KubeSecret != "" {
+		if err := storeDeviceID(ctx, cfg.KubeSecret, string(st.Self.ID)); err != nil {
+			log.Fatalf("storing device ID in kube secret: %v", err)
+		}
+		if cfg.AuthOnce {
+			// We were told to only auth once, so any secret-bound
+			// authkey is no longer needed. We don't strictly need to
+			// wipe it, but it's good hygiene.
+			log.Printf("Deleting authkey from kube secret")
+			if err := deleteAuthKey(ctx, cfg.KubeSecret); err != nil {
+				log.Fatalf("deleting authkey from kube secret: %v", err)
+			}
+		}
+	}
+
+	log.Println("Startup complete, waiting for shutdown signal")
+	// Sleep for ever, tailscaled is handled from a goroutine in
+	// startAndAuthTailscaled.
+	select {}
+}
+
+// startAndAuthTailscaled starts the tailscale daemon and attempts to
+// auth it, according to the settings in cfg. If successful, returns a
+// tailscale.LocalClient connected to the running tailscaled.
+func startAndAuthTailscaled(ctx context.Context, cfg *settings) (*ipnstate.Status, error) {
+	args := tailscaledArgs(cfg)
+	sigCh := make(chan os.Signal, 1)
+	signal.Notify(sigCh, unix.SIGTERM, unix.SIGINT)
+	// tailscaled runs without context, since it needs to persist
+	// beyond the startup timeout in ctx.
+	cmd := exec.Command("tailscaled", args...)
+	cmd.Stdout = os.Stdout
+	cmd.Stderr = os.Stderr
+	log.Printf("Starting tailscaled")
+	if err := cmd.Start(); err != nil {
+		return nil, fmt.Errorf("starting tailscaled failed: %v", err)
+	}
+	daemonPid := cmd.Process.Pid
+	go func() {
+		<-sigCh
+		log.Printf("Received SIGTERM from container runtime, shutting down tailscaled")
+		cmd.Process.Signal(unix.SIGTERM)
+	}()
+	go func() {
+		// Reap all processes, since we are PID1 and need to collect
+		// zombies.
+		for {
+			var status unix.WaitStatus
+			pid, err := unix.Wait4(-1, &status, 0, nil)
+			if err != nil {
```

Comment:
I think you probably need to handle `unix.EINTR` (and just continue) here?

Reply:
Done